File: 프로그래머스/Level2/빛의 경로 사이클.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool visit[501][501][4];
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

int getCycleLen(vector<string>& grid, int x, int y, int dir, int len){
    if(visit[y][x][dir] == true) return len;
    
    visit[y][x][dir] = true;
    int next_dir;
    if(grid[y][x] == 'S'){
        next_dir = dir;
    }
    else if(grid[y][x] == 'L'){
        if(dir == 0) next_dir = 2;
        else if(dir == 1) next_dir = 3;
        else if(dir == 2) next_dir = 1;
        else next_dir = 0;
    }
    else{
        if(dir == 0) next_dir = 3;
        else if(dir == 1) next_dir = 2;
        else if(dir == 2) next_dir = 0;
        else next_dir = 1;
    }
    
    int next_x = x + dx[next_dir];
    int next_y = y + dy[next_dir];
    if(next_x == grid[0].size()) next_x = 0;
    if(next_x < 0) next_x = grid[0].size() - 1;
    if(next_y == grid.size()) next_y = 0;
    if(next_y < 0) next_y = grid.size() - 1;
    
    return getCycleLen(grid, next_x, next_y, next_dir, len + 1);
}
vector<int> solution(vector<string> grid) {
    vector<int> answer;
    
    for(int i=0; i<grid.size(); i++){
        for(int j=0; j<grid[i].size(); j++){
            for(int k=0; k<4; k++){
                int cycle_len = getCycleLen(grid, j, i, k, 0);
                if(cycle_len > 0) answer.push_back(cycle_len);
            }
        }
    }
    sort(answer.begin(), answer.end());
    return answer;
}

File: 프로그래머스/Level2/빛의 경로 사이클.java
import java.util.*;

class Solution {

    //오른쪽, 아래, 왼쪽, 위
    private int[] dr = {0, 1, 0, -1};
    private int[] dc = {1, 0, -1, 0};

    private int rowSize;
    private int colSize;
    
    public int[] solution(String[] grid) {

        rowSize = grid.length;
        colSize = grid[0].length();
        
        boolean[][][] map = new boolean[rowSize][colSize][4];
        List<Integer> answer = new ArrayList<>();

        for (int r = 0; r < rowSize; ++r) {
            for (int c = 0; c < colSize; ++c) {
                for (int d = 0; d < 4; ++d) {
                    if (!map[r][c][d]) {
                        int cycleLen = travel(map, grid, r, c, d);
                        answer.add(cycleLen);
                    }
                }
            }
        }

        Collections.sort(answer);
        return answer.stream().mapToInt(i -> i).toArray();
    }

    private int travel(boolean[][][] map, String[] grid, int r, int c, int d) {
        int cnt = 0;
        while(!map[r][c][d]) {
            map[r][c][d] = true;
            
            if (grid[r].charAt(c) == 'R') {
                d = (d + 1) % 4;
            }
            if (grid[r].charAt(c) == 'L') {
                d = (d + 3) % 4;
            }
            
            r = r + dr[d];
            c = c + dc[d];
            
            //격자의 끝을 넘어가는 경우
            if (r == -1) {
                r = rowSize - 1;
            }
            if (r == rowSize) {
                r = 0;
            }
            if (c == -1) {
                c = colSize - 1;
            }
            if (c == colSize) {
                c = 0;
            }
            
            ++cnt;
        }
        return cnt;
    }
}
